### PR TITLE
Engine

### PR DESCRIPTION
**Pull request by jk2227**
https://github.com/cbora/predictionio-template-java-universal-recommender/issues/2

### CODE REVIEW

**Comment by jk2227** [comment]
Current progress is blocked by the other pull requests that are currently out as the Algorithm class is huge + depends on other classes (e.g. PopModel, PreparedData)

**Comment by jk2227** [comment]
@dszeto it would be great if we could chat about how mahout to build; I sent more information on Slack about it but this seems to be our blocker to finally merging in something that builds properly

@cbora I made a lot of edits to your part; if you can take a look at SelfCleaningDataSource trait and fill in all of the methods that are supposed to be implemented by it in DataSource.java in a separate branch that would be great

@cfifty also made slight edits to your part; feel free to fetch from my repo and make necessary edits

File: src/main/java/org/template/recommendation/Preparator.java [hunk 1]
```diff
@@ -1,12 +1,82 @@
...
+            // and append any new ids, so after all are constructed we have all user ids in the last dictionary
+            IndexedDatasetSpark ids = IndexedDatasetSpark.apply(eventIDS.rdd(), userDictionary, sc);
+            userDictionary = scala.Option.apply(ids.rowIDs());
+
+            // append the transformation to the indexedDatasets list
+            indexedDatasets.add(new Tuple2<>(eventName, ids));
+        }
+
+        // now make sure all matrices have identical row space since this corresponds to all users
+        List<Tuple2<String,IndexedDataset>> rowAdjustedIds = userDictionary.map(userDict  -> {
```
Comment (cfifty) [inline src/main/java/org/template/recommendation/Preparator.java]: fix

Comment (cfifty) [inline src/main/java/org/template/recommendation/Preparator.java]: Will make a new branch, pull from yours, and merge in my code from Preparator.  You can pull from there.

**Comment by cfifty** [comment]
Just incorporated the Preparator files into jk2227's engine files and put the combination in Preparator branch.  I tried incorporating Project Lombok in Preparator to reduce code lines, but ran into trouble with using Lombok's getter with final variables and getting variables using methods generated by Lombok's Getter

File: src/main/java/org/template/recommendation/Preparator.java (hunk 1 above)
Comment (jk2227) [inline src/main/java/org/template/recommendation/Preparator.java]: https://github.com/pferrel/template-scala-parallel-universal-recommendation/blob/master/src/main/scala/Preparator.scala 

Refer to lines 51-58 @cfifty

Comment (cfifty) [inline src/main/java/org/template/recommendation/Preparator.java]: Looks like I was converting an older version of preparator: 

https://github.com/PredictionIO/template-scala-parallel-universal-recommendation/blob/master/src/main/scala/Preparator.scala

Lines 54-56 @jk2227

Comment (cfifty) [inline src/main/java/org/template/recommendation/Preparator.java]: Should we use your link (the parallel recommender) or the actionml universe recommender (https://github.com/cfifty/universal-recommender/blob/master/src/main/scala/Preparator.scala)?

**Comment by jk2227** [comment]
@dszeto

**Pull request merged**